src/ipips/ipip-0431.md

```diff
@@ -0,0 +1,197 @@
+ - Instead, we treat the response body as a new container format combining the CARv1 file with additional data.
+ - Clients must explicitly request this new container format. Existing clients not aware of the new metadata will not receive responses in the new format.
+
+#### Denial of Service attacks
+
+Computing the signature for the metadata blcok has a non-negligible performance cost. To mittigate DoS attacks, we designed the metadata to be highly cacheable. When a gateway receives two requests for the same content, it can return the same metadata block in both responses, including the signature. This allows gateway operators to deploy a traditional caching layer operating at the HTTP protocol, the cache does not need to understand any specifics of IPFS and Trustless Gateway protocols.
+
+### Alternatives
+
+#### HTTP Trailers
+
+Instead of adding a new content type argument, we were considering sending the additional metadata
+in HTTP response trailers. Unfortunately, HTTP trailers are not widely supported by the ecosystem.
+Nginx proxy module discards them, [browser `Fetch API` does not allow JS clients to access trailer
+headers](https://github.com/mdn/browser-compat-data/issues/14703), neither does the Rust `reqwest` client.
+
+#### New Content-Type
+
+We could introduce a new content type that is not CARv3, but a thin envelope
+around CARv1 with purpose of streaming over HTTP (e.g. `Content-Type:
+application/vnd.ipld.car-stream`).
+
+It would have three fields:
+- `car-stream-header` (optional DAG-CBOR)
+- `car` (same as `application/vnd.ipld.car;version=1`)
+- `car-stream-end` (optional DAG-CBOR)
+
+This will be enough to append DAG-CBOR  manifest at the end of the stream. It
+would be effectively the same CAR byte stream, but with different
+`Content-Type`.
+
+Upside of this solution:
+
+- does not require registering new codec, or mixing data plane with control
+  plane, no sniffing the last DAG-CBOR block
+
+Downsides of this solution:
+
+- maintenance cost, requires duplicating of all CAR-related tests and features
+- ecosystem opportunity cost, in creating new content type, we increase
+  cognitive overhead for everyone working with IPFS over HTTP
+- no backward-compatible interop with existing tools and gateways that only
+  speak `application/vnd.ipld.car`
+- distracts us away from working on things like large blocks and CARv3
+
+#### Create CARv3
+
+We could admit we've clearly hit limitation of what we can do with HTTP and CARv1 and CARv2 and stop abusing existing CARv1 by mixing data plane with control plane.
+
+Spend energy on creating CARv3 that solves the problems from "Motivation" section and more:
+- optional index or key-value metadata before or after data
+- native truncation detection and standardized error handling and passing during streaming
+- support for things like [Large Blocks](https://discuss.ipfs.tech/t/supporting-large-ipld-blocks/15093/)
+
+TODO: link to some public artifact about CARv3
```

Comment:
Flagging this TODO to show in the PR discussion.

Any suggestions for the artefacts I can link to?

Reply:
@aschmahmann do we have anything on GH?